tests/test_runner/tests.py

```diff
@@ -468,3 +469,25 @@ def test_run_checks_passes_and_teardown_raises(self):
                     runner.run_tests(['test_runner_apps.sample.tests_sample.TestDjangoTestCase'])
             self.assertTrue(teardown_databases.called)
             self.assertFalse(teardown_test_environment.called)
+
+class CircularRefTestRunnerTest(unittest.TestCase):
+    # This is a regression for #26552 and part of #31051
+    # This test lets the DiscoveryRunner go through the entire process of running tests, in particular setting up the
+    # database and handling serialized_rollback between tests. The models contain circular references with natural
+    # keys, to check that this is not a problem for the serialization code anymore. The migrations add instances of
+    # these models, to check that deserializing such instances with forward references is not problematic anymore.
+    def test_setup_databases(self):
+        with modify_settings(INSTALLED_APPS={'append': 'test_runner_apps.circular_ref'}):
+            from test_runner_apps.circular_ref.models import CircularRefThingA, CircularRefThingB, SelfRefThing
```

Comment:
`test_runner_apps` should likely live in `test_runner.app_something` instead. Also this import should be moved to the top.

Reply:
`test_runner_apps` already exists, so moving that seems out of scope for this PR.

IIRC the import cannot be at the top, since then a model is imported that is not in `INSTALLED_APPS`, which I think caused a problem. I'll doublecheck and add a comment.

Reply:
Turns out the import was actually unused, so I removed it and will resolve this conversation. If you want to continue discussing moving `test_runner_apps`, please add a new comment to the PR for that.